File: src/main/java/net/ccbluex/liquidbounce/ui/client/clickgui/style/styles/nlclickgui/NeverloseGui.java
package net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui;

import com.mojang.realmsclient.gui.ChatFormatting;
import net.ccbluex.liquidbounce.FDPClient;
import net.ccbluex.liquidbounce.features.module.Category;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.fdpdropdown.SideGui.SideGui;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Config.Configs;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Config.NeverloseConfigManager;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Settings.BoolSetting;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Settings.ColorSetting;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Settings.Numbersetting;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Settings.StringsSetting;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.Animation;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.Direction;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.impl.DecelerateAnimation;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.impl.EaseInOutQuad;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.impl.SmoothStepAnimation;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.blur.BloomUtil;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.blur.GaussianBlur;
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.round.RoundedUtil;
import net.ccbluex.liquidbounce.ui.font.Fonts;
import net.ccbluex.liquidbounce.ui.font.fontmanager.api.FontRenderer;
import net.ccbluex.liquidbounce.ui.client.gui.GuiSpotify;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.client.shader.Framebuffer;
import net.minecraft.util.ChatAllowedCharacters;
import net.minecraft.util.ResourceLocation;
import org.lwjgl.opengl.GL11;

import java.awt.*;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class NeverloseGui extends GuiScreen {

    public static NeverloseGui INSTANCE;

    public int x,y,w,h;

    public Animation alphaani;

    public static Color neverlosecolor = new Color(28,133,192);

    public NlSub selectedSub = null;

    public List<NlTab> nlTabs = new ArrayList<>();

    public boolean Loader = true;

    private final SideGui sideGui = new SideGui();

    private int x2;
    private int y2;

    private boolean dragging,settings,search;
    private String searchText = "";

    private final ResourceLocation defaultAvatar = new ResourceLocation(FDPClient.CLIENT_NAME.toLowerCase() + "/64.png");
    private final ResourceLocation spotifyIcon = new ResourceLocation(FDPClient.CLIENT_NAME.toLowerCase() + "/spotify.png");
    private ResourceLocation avatarTexture = defaultAvatar;
    private boolean avatarLoaded;

    private NlSetting nlSetting;

    private int spotifyX;
    private int spotifyY;
    private final int spotifySize = 16;

    private Animation searchanim = new EaseInOutQuad(400,1, Direction.BACKWARDS);


    public Configs configs = new Configs();
    private final NeverloseConfigManager configManager = new NeverloseConfigManager();

    public NeverloseGui(){
        INSTANCE = this;
        x = 100;
        y = 100;
        w = 500;
        h = 380;

        int y2 = 0;
        int u2 =0;
        List<Category> orderedCategories = new ArrayList<>();
        orderedCategories.add(Category.CLIENT);

        for (Category type : Category.values()) {
            if (!orderedCategories.contains(type)) {
                orderedCategories.add(type);
            }
        }

        for (Category type : orderedCategories){
            if(type.name().equalsIgnoreCase("World") ||
                    type.name().equalsIgnoreCase("Interface")) continue;

            nlTabs.add(new NlTab(type,u2 + y2 + 40));

            for (Category.SubCategory subCategory: type.getSubCategories()){
                u2 += 17;
            }

            y2 += 14;
        }

        nlSetting = new NlSetting();

    }


    @Override
    public void initGui() {
        super.initGui();
        configManager.refresh();

        alphaani = new EaseInOutQuad(300,0.6,Direction.FORWARDS);
        sideGui.initGui();
    }

    private Framebuffer bloomFramebuffer = new Framebuffer(1, 1, false);

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {

        GL11.glPushMatrix();





        if (Loader && !nlTabs.isEmpty()){
            selectedSub = nlTabs.get(0).nlSubList.get(0);
            Loader = false;
        }

        if(dragging) {
            this.x = x2 + mouseX;
            this.y = y2 + mouseY;
        }

        bloomFramebuffer = RenderUtil.createFrameBuffer(bloomFramebuffer);

        bloomFramebuffer.framebufferClear();
        bloomFramebuffer.bindFramebuffer(true);
        RoundedUtil.drawRound(x,y,w,h,4,nlSetting.Light ? new Color(240,245,248,230) : new Color(7,13,23,230));
        bloomFramebuffer.unbindFramebuffer();

        BloomUtil.renderBlur(bloomFramebuffer.framebufferTexture, 6, 3);

        StencilUtil.initStencilToWrite();
        RoundedUtil.drawRound(x,y,w,h,4,nlSetting.Light ? new Color(240,245,248,230) : new Color(7,13,23,230));
        StencilUtil.readStencilBuffer(1);
        GaussianBlur.renderBlur(/* 大小 */10);
        StencilUtil.uninitStencilBuffer();

        RoundedUtil.drawRound(x,y,w,h,2,nlSetting.Light ? new Color(240,245,248,230) : new Color(7,13,23,230));

        RoundedUtil.drawRound(x + 90,getContentTop(),w - 90,getContentHeight(),1,nlSetting.Light ? new Color(255,255,255) :new Color(0,0,0,235));

        RoundedUtil.drawRound(x + 90,y,w - 90,getContentTopOffset() , 1,nlSetting.Light ?new Color(255,255,255)  :new Color(13,13,11));

        RoundedUtil.drawRound(x + 90,y + 39,w - 90,1 , 0,nlSetting.Light ? new Color(213,213,213) : new Color(26,26,26));

        RoundedUtil.drawRound(x + 89 ,y,1,h,0,nlSetting.Light ? new Color(213,213,213) : new Color(26,26,26));

        GL11.glEnable(GL11.GL_BLEND);

        ensureAvatarTexture();
        mc.getTextureManager().bindTexture(avatarTexture);

        int footerLineY = y + h - 35;
        int avatarY = footerLineY + 9;

        RoundedUtil.drawRoundTextured(x + 4 , avatarY,20,20,10f,1);

        Fonts.Nl_18.drawString(mc.getSession().getUsername(),x + 29 , avatarY + 1,nlSetting.Light ? new Color(51,51,51).getRGB() : -1);

        Fonts.Nl_16.drawString(ChatFormatting.GRAY + "Till: " + ChatFormatting.RESET + new SimpleDateFormat("dd:MM").format(new Date()) + " " + new SimpleDateFormat("HH:mm").format(new Date()),x + 29 , avatarY + 13,neverlosecolor.getRGB());

        if (!nlSetting.Light) {
            NLOutline("FDPCLIENT", Fonts.NLBold_28, x + 7, y + 12, -1, neverlosecolor.getRGB(), 0.7f);

        }else {
            Fonts.NLBold_28.drawString("FDP", x + 8, y + 12, new Color(51,51,51).getRGB(), false);
        }

        RoundedUtil.drawRound(x , footerLineY,89,1 , 0,nlSetting.Light ? new Color(213,213,213) :new Color(26,26,26));

        spotifyX = x + 165;
        spotifyY = y + 14;
        RenderUtil.drawImage(spotifyIcon, spotifyX, spotifyY, spotifySize, spotifySize);

        for (NlTab nlTab : nlTabs){
            nlTab.x = x;
            nlTab.y = y;
            nlTab.w = w;
            nlTab.h = h;

            nlTab.draw(mouseX,mouseY);
        }

        Fonts.NlIcon.nlfont_20.getNlfont_20().drawString("x", (float) (x + w - 50 + (search || !searchanim.isDone() ? -83 * searchanim.getOutput() : 0)),y + 17,settings ? neverlosecolor.getRGB() : NeverloseGui.INSTANCE.getLight() ? new Color(95,95,95).getRGB() :-1);

        Fonts.NlIcon.nlfont_20.getNlfont_20().drawString("j",x + w - 30,y + 18,search ? neverlosecolor.getRGB() :NeverloseGui.INSTANCE.getLight() ? new Color(95,95,95).getRGB() : -1);

        searchanim.setDirection(search ? Direction.FORWARDS : Direction.BACKWARDS);

        if (search || !searchanim.isDone()){
            RenderUtil.drawRoundedRect((float) (x + w - 30 -(85 * searchanim.getOutput() )), (float) (y + 12), (float) (80 * searchanim.getOutput()),15,1,new Color(3,13,26).getRGB(),1,NeverloseGui.INSTANCE.getLight() ? new Color(95,95,95).getRGB() : new Color(28,133,192).getRGB());
            Fonts.Nl_16.drawString(searchText, (float) (x + w - 26 -(85 * searchanim.getOutput() )), y + 15, NeverloseGui.INSTANCE.getLight() ? new Color(95,95,95).getRGB() : -1);
        }

        if (settings){
            nlSetting.draw(mouseX,mouseY);
        }

        RoundedUtil.drawRoundOutline(x + 105,y+10,45 + 10,16 + 5,2,0.1f,NeverloseGui.INSTANCE.getLight() ? new Color(245,245,245) : new Color(13,13,11),RenderUtil.isHovering(x + 105, y + 10, 45 + 10, 16 + 5, mouseX, mouseY) ? neverlosecolor :new Color(19,19,17));

        Fonts.Nl_18.drawString( "Save",x + 128,y+18 ,NeverloseGui.INSTANCE.getLight()? new Color(18,18,19).getRGB() : -1);

        Fonts.NlIcon.nlfont_20.getNlfont_20().drawString("K",x + 110,y+19, NeverloseGui.INSTANCE.getLight()? new Color(18,18,19).getRGB() : -1);

        GL11.glPopMatrix();
        sideGui.drawScreen(mouseX, mouseY, partialTicks, 255);
        super.drawScreen(mouseX, mouseY, partialTicks);
    }

    private void ensureAvatarTexture() {
        if (!avatarLoaded) {
            avatarTexture = defaultAvatar;
            avatarLoaded = true;
        }
    }

    public static void NLOutline(String str, FontRenderer fontRenderer, float x, float y, int color, int color2, float size) {
        fontRenderer.drawString(str, x + size, y, color2, false);
        fontRenderer.drawString(str, x, y - size, color2, false);
        fontRenderer.drawString(str, x, y, color, false);
    }


    @Override
    protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException {
        boolean wasFocused = sideGui.getFocused();
        sideGui.mouseClicked(mouseX, mouseY, mouseButton);

        if (!wasFocused) {
            nlTabs.forEach(e -> e.click(mouseX,mouseY,mouseButton));
            if (settings){
                nlSetting.click(mouseX,mouseY,mouseButton);
            }
            if (mouseButton ==0){
                if (RenderUtil.isHovering(spotifyX, spotifyY, spotifySize, spotifySize, mouseX, mouseY)) {
                    mc.displayGuiScreen(new GuiSpotify(this));
                    return;
                }

                if(RenderUtil.isHovering(x + 110,y,w - 110,h - 300 ,mouseX,mouseY)) {
                    this.x2 = (int) (x - mouseX);
                    this.y2 = (int) (y - mouseY);
                    this.dragging = true;
                }

                if (RenderUtil.isHovering(x + 105, y + 10, 45 + 10, 16 + 5, mouseX, mouseY)) {
                    if (configManager.getActiveConfig() != null) {
                        configManager.saveConfig(configManager.getActiveConfig().getName());
                    } else {
                        FDPClient.fileManager.saveAllConfigs();
                        configManager.refresh();
                    }
                }

                if (RenderUtil.isHovering((float) (x + w - 50+ (search || !searchanim.isDone() ? -83 * searchanim.getOutput() : 0)),y + 17,Fonts.NlIcon.nlfont_24.getNlfont_24().stringWidth("x"),Fonts.NlIcon.nlfont_24.getNlfont_24().getHeight(),mouseX,mouseY)){
                    settings = !settings;
                    dragging = false;
                    nlSetting.x = this.x + this.w + 20;
                    nlSetting.y = this.y;
                }

                if (RenderUtil.isHovering(x + w - 30,y + 18,Fonts.NlIcon.nlfont_20.getNlfont_20().stringWidth("j"),Fonts.NlIcon.nlfont_20.getNlfont_20().getHeight(),mouseX,mouseY)){
                    search = !search;
                    dragging = false;
                    if (!search) {
                        searchText = "";
                    }
                }

            }
        }
        super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    protected void mouseReleased(int mouseX, int mouseY, int state) {
        boolean wasFocused = sideGui.getFocused();
        sideGui.mouseReleased(mouseX, mouseY, state);

        if (!wasFocused) {
            nlTabs.forEach( e -> e.released(mouseX,mouseY,state));

            if(state == 0) {
                this.dragging = false;
            }


            if (settings){
                nlSetting.released(mouseX,mouseY,state);
            }
        }

        super.mouseReleased(mouseX, mouseY, state);
    }

    @Override
    protected void keyTyped(char typedChar, int keyCode) throws IOException {
        sideGui.keyTyped(typedChar, keyCode);

        if (search) {
            if (keyCode == 1) {
                search = false;
                searchText = "";
                return;
            }
            if (keyCode == 14) {
                if (!searchText.isEmpty()) {
                    searchText = searchText.substring(0, searchText.length() - 1);
                }
                return;
            }
            if (ChatAllowedCharacters.isAllowedCharacter(typedChar)) {
                searchText = searchText + typedChar;
                return;
            }
        }

        nlTabs.forEach( e -> e.keyTyped(typedChar,keyCode));
        super.keyTyped(typedChar, keyCode);
    }

    public boolean isSearching() {
        return search && !searchText.isEmpty();
    }

    public String getSearchText() {
        return searchText;
    }


    public int getContentTopOffset() {
        return Math.max(70, h - 290);
    }

    public int getContentTop() {
        return y + getContentTopOffset();
    }

    public int getContentHeight() {
        return h - getContentTopOffset() - 40;
    }


    public static NeverloseGui getInstance() {
        return INSTANCE;
    }

    public NeverloseConfigManager getConfigManager() {
        return configManager;
    }

    public boolean getLight(){
        return nlSetting.Light;
    }

}

File: src/main/java/net/ccbluex/liquidbounce/ui/client/clickgui/style/styles/nlclickgui/NeverloseGui.kt
package net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui

import com.mojang.realmsclient.gui.ChatFormatting
import net.ccbluex.liquidbounce.FDPClient
import net.ccbluex.liquidbounce.features.module.Category
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Config.Configs
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Config.NeverloseConfigManager
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Settings.BoolSetting
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Settings.ColorSetting
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Settings.Numbersetting
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.Settings.StringsSetting
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.Animation
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.Direction
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.impl.DecelerateAnimation
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.impl.EaseInOutQuad
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.animations.impl.SmoothStepAnimation
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.blur.BloomUtil
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.blur.GaussianBlur
import net.ccbluex.liquidbounce.ui.client.clickgui.style.styles.nlclickgui.round.RoundedUtil
import net.ccbluex.liquidbounce.ui.font.Fonts
import net.ccbluex.liquidbounce.ui.font.fontmanager.api.FontRenderer
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiScreen
import net.minecraft.client.renderer.texture.DynamicTexture
import net.minecraft.client.shader.Framebuffer
import net.minecraft.util.ResourceLocation
import org.lwjgl.opengl.GL11
import java.awt.Color
import java.io.IOException
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date
import javax.imageio.ImageIO

class NeverloseGui : GuiScreen() {

    var x = 0
    var y = 0
    var w = 0
    var h = 0

    lateinit var alphaani: Animation

    var subCategory: Category.SubCategory? = null

    val nlTabs: MutableList<NlTab> = ArrayList()

    var loader = true

    private var x2 = 0
    private var y2 = 0

    private var dragging = false
    private var settings = false
    private var search = false
    private var head = true

    private val nlSetting = NlSetting()

    private val searchanim: Animation = EaseInOutQuad(400.0, 1.0, Direction.BACKWARDS)

    var configs = Configs()
    private val configManager = NeverloseConfigManager()

    private var bloomFramebuffer = Framebuffer(1, 1, false)

    init {
        INSTANCE = this
        x = 100
        y = 100
        w = 430
        h = 300

        var y2 = 0
        var u2 = 0
        for (type in Category.values()) {
            if (type.name.equals("World", ignoreCase = true) ||
                type.name.equals("Interface", ignoreCase = true)
            ) continue

            nlTabs.add(NlTab(type, u2 + y2 + 40))

            for (subCategory in type.subCategories) {
                u2 += 17
            }

            y2 += 14
        }

        if (head) {
            try {
                Minecraft.getMinecraft().textureManager.loadTexture(
                    ResourceLocation("nb"),
                    DynamicTexture(ImageIO.read(URL("https://q.qlogo.cn/headimg_dl?dst_uin=" + "2165728490" + "&spec=100"))))
                head = false
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }

    override fun initGui() {
        super.initGui()
        configManager.refresh()

        alphaani = EaseInOutQuad(300.0, 0.6, Direction.FORWARDS)
    }

    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        GL11.glPushMatrix()

        if (loader && nlTabs.isNotEmpty()) {
            subCategory = nlTabs[0].nlSubList[0].subCategory
            loader = false
        }

        if (dragging) {
            x = x2 + mouseX
            y = y2 + mouseY
        }

        bloomFramebuffer = RenderUtil.createFrameBuffer(bloomFramebuffer)

        bloomFramebuffer.framebufferClear()
        bloomFramebuffer.bindFramebuffer(true)
        RoundedUtil.drawRound(x, y, w, h, 4f, if (nlSetting.Light) Color(240, 245, 248, 230) else Color(7, 13, 23, 230))
        bloomFramebuffer.unbindFramebuffer()

        BloomUtil.renderBlur(bloomFramebuffer.framebufferTexture, 6, 3)

        StencilUtil.initStencilToWrite()
        RoundedUtil.drawRound(x, y, w, h, 4f, if (nlSetting.Light) Color(240, 245, 248, 230) else Color(7, 13, 23, 230))
        StencilUtil.readStencilBuffer(1)
        GaussianBlur.renderBlur(10)
        StencilUtil.uninitStencilBuffer()

        RoundedUtil.drawRound(x, y, w, h, 2f, if (nlSetting.Light) Color(240, 245, 248, 230) else Color(7, 13, 23, 230))
        RoundedUtil.drawRound(x + 90, y + 40, w - 90, h - 40, 1f, if (nlSetting.Light) Color(255, 255, 255) else Color(9, 9, 9))
        RoundedUtil.drawRound(x + 90, y, w - 90, h - 260, 1f, if (nlSetting.Light) Color(255, 255, 255) else Color(13, 13, 11))

        RoundedUtil.drawRound(x + 90, y + 39, w - 90, 1f, 0f, if (nlSetting.Light) Color(213, 213, 213) else Color(26, 26, 26))
        RoundedUtil.drawRound(x + 89, y, 1, h, 0f, if (nlSetting.Light) Color(213, 213, 213) else Color(26, 26, 26))

        GL11.glEnable(GL11.GL_BLEND)

        mc.textureManager.bindTexture(ResourceLocation("nb"))

        RoundedUtil.drawRoundTextured(x + 4, y + 274, 20, 20, 10f, 1f)

        Fonts.Nl_18.drawString(mc.session.username, x + 29, y + 275, if (nlSetting.Light) Color(51, 51, 51).rgb else -1)

        Fonts.Nl_16.drawString(
            ChatFormatting.GRAY.toString() + "Till: " + ChatFormatting.RESET + SimpleDateFormat("dd:MM").format(Date()) + " " + SimpleDateFormat(
                "HH:mm"
            ).format(Date()),
            x + 29,
            y + 287,
            neverlosecolor.rgb
        )

        if (!nlSetting.Light) {
            NLOutline("FDPCLIENT", Fonts.NLBold_28, (x + 7).toFloat(), (y + 12).toFloat(), -1, neverlosecolor.rgb, 0.7f)
        } else {
            Fonts.NLBold_28.drawString("FDP", (x + 8).toFloat(), (y + 12).toFloat(), Color(51, 51, 51).rgb, false)
        }

        RoundedUtil.drawRound(x, y + 265, 89, 1, 0f, if (nlSetting.Light) Color(213, 213, 213) else Color(26, 26, 26))

        for (nlTab in nlTabs) {
            nlTab.x = x
            nlTab.y = y
            nlTab.w = w
            nlTab.h = h

            nlTab.draw(mouseX, mouseY)
        }

        Fonts.NlIcon.nlfont_20.getNlfont_20().drawString(
            "x",
            (x + w - 50 + if (search || !searchanim.isDone) (-83 * searchanim.output).toFloat() else 0f).toFloat(),
            (y + 17).toFloat(),
            if (settings) neverlosecolor.rgb else if (INSTANCE.light) Color(95, 95, 95).rgb else -1
        )

        Fonts.NlIcon.nlfont_20.getNlfont_20().drawString(
            "j",
            (x + w - 30).toFloat(),
            (y + 18).toFloat(),
            if (search) neverlosecolor.rgb else if (INSTANCE.light) Color(95, 95, 95).rgb else -1
        )

        searchanim.setDirection(if (search) Direction.FORWARDS else Direction.BACKWARDS)

        if (search || !searchanim.isDone) {
            RenderUtil.drawRoundedRect(
                (x + w - 30 - 85 * searchanim.output).toFloat(),
                (y + 12).toFloat(),
                (80 * searchanim.output).toFloat(),
                15f,
                1,
                Color(3, 13, 26).rgb,
                1,
                if (INSTANCE.light) Color(95, 95, 95).rgb else Color(28, 133, 192).rgb
            )
        }

        if (settings) {
            nlSetting.draw(mouseX, mouseY)
        }

        RoundedUtil.drawRoundOutline(
            x + 105,
            y + 10,
            55,
            21,
            2f,
            0.1f,
            if (INSTANCE.light) Color(245, 245, 245) else Color(13, 13, 11),
            if (RenderUtil.isHovering(x + 105, y + 10, 55, 21, mouseX, mouseY)) neverlosecolor else Color(19, 19, 17)
        )

        Fonts.Nl_18.drawString("Save", (x + 128).toFloat(), (y + 18).toFloat(), if (INSTANCE.light) Color(18, 18, 19).rgb else -1)

        Fonts.NlIcon.nlfont_20.getNlfont_20().drawString("K", (x + 110).toFloat(), (y + 19).toFloat(), if (INSTANCE.light) Color(18, 18, 19).rgb else -1)

        GL11.glPopMatrix()
        super.drawScreen(mouseX, mouseY, partialTicks)
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        nlTabs.forEach { it.click(mouseX, mouseY, mouseButton) }
        if (settings) {
            nlSetting.click(mouseX, mouseY, mouseButton)
        }
        if (mouseButton == 0) {
            if (RenderUtil.isHovering(x + 110, y, w - 110, h - 260, mouseX, mouseY)) {
                x2 = (x - mouseX)
                y2 = (y - mouseY)
                dragging = true
            }

            if (RenderUtil.isHovering(x + 105, y + 10, 55, 21, mouseX, mouseY)) {
                if (configManager.activeConfig != null) {
                    configManager.saveConfig(configManager.activeConfig!!.name)
                } else {
                    FDPClient.fileManager.saveAllConfigs()
                    configManager.refresh()
                }
            }

            if (RenderUtil.isHovering(
                    (x + w - 50 + if (search || !searchanim.isDone) -83 * searchanim.output else 0.0).toFloat(),
                    (y + 17).toFloat(),
                    Fonts.NlIcon.nlfont_24.getNlfont_24().stringWidth("x"),
                    Fonts.NlIcon.nlfont_24.getNlfont_24().height,
                    mouseX,
                    mouseY
                )
            ) {
                settings = !settings
                dragging = false
                nlSetting.x = x + w + 20
                nlSetting.y = y
            }

            if (RenderUtil.isHovering(
                    (x + w - 30).toFloat(),
                    (y + 18).toFloat(),
                    Fonts.NlIcon.nlfont_20.getNlfont_20().stringWidth("j"),
                    Fonts.NlIcon.nlfont_20.getNlfont_20().height,
                    mouseX,
                    mouseY
                )
            ) {
                search = !search
                dragging = false
            }
        }
        super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, state: Int) {
        nlTabs.forEach { it.released(mouseX, mouseY, state) }

        if (state == 0) {
            dragging = false
        }

        if (settings) {
            nlSetting.released(mouseX, mouseY, state)
        }

        super.mouseReleased(mouseX, mouseY, state)
    }

    override fun keyTyped(typedChar: Char, keyCode: Int) {
        nlTabs.forEach { it.keyTyped(typedChar, keyCode) }
        super.keyTyped(typedChar, keyCode)
    }

    val light: Boolean
        get() = nlSetting.Light

    companion object {
        @JvmStatic
        lateinit var INSTANCE: NeverloseGui

        @JvmStatic
        val neverlosecolor: Color = Color(28, 133, 192)

        @JvmStatic
        fun getInstance(): NeverloseGui = INSTANCE

        @JvmStatic
        fun NLOutline(str: String, fontRenderer: FontRenderer, x: Float, y: Float, color: Int, color2: Int, size: Float) {
            fontRenderer.drawString(str, x + size, y, color2, false)
            fontRenderer.drawString(str, x, y - size, color2, false)
            fontRenderer.drawString(str, x, y, color, false)
        }
    }
}
